File: tests/ClaudeToolbar.Core.Tests/Sessions/HostChainTests.cs
using ClaudeToolbar.Core.Sessions;

namespace ClaudeToolbar.Core.Tests.Sessions;

public class HostChainTests
{
    private static readonly DateTimeOffset T0 = new(2026, 9, 8, 9, 0, 0, TimeSpan.Zero);

    // startOffsetSeconds: larger = started later. Children start after their parents.
    private static ProcessRecord P(int pid, int parent, string name, bool window = false, int startOffsetSeconds = 0) =>
        new(pid, parent, name, T0.AddSeconds(startOffsetSeconds), window);

    private static IReadOnlyDictionary<int, ProcessRecord> Table(params ProcessRecord[] records) => records.ToDictionary(r => r.Pid);

    [Fact]
    public void FindsWindowsTerminalThreeLevelsUp()
    {
        var table = Table(
            P(100, 90, "claude", startOffsetSeconds: 30),
            P(90, 80, "pwsh", startOffsetSeconds: 20),
            P(80, 70, "openconsole", startOffsetSeconds: 10),
            P(70, 1, "windowsterminal", window: true));
        Assert.Equal(new HostChainResult(70, "Windows Terminal", true), HostChain.Resolve(100, table, KnownHosts.Windows));
    }

    [Fact]
    public void TheClaudeCliIsNotTheDesktopApp()
    {
        var table = Table(P(100, 90, "claude", startOffsetSeconds: 10), P(90, 1, "windowsterminal", window: true));
        Assert.Equal("Windows Terminal", HostChain.Resolve(100, table, KnownHosts.Windows)!.DisplayName);
    }

    [Fact]
    public void TheDesktopAppCountsWhenItOwnsAWindow()
    {
        var table = Table(P(100, 50, "claude", startOffsetSeconds: 10), P(50, 1, "claude", window: true));
        Assert.Equal(new HostChainResult(50, "Claude", true), HostChain.Resolve(100, table, KnownHosts.Windows));
    }

    [Fact]
    public void VsCodeHelperWithoutAWindowIsSkippedForItsMainProcess()
    {
        var table = Table(
            P(100, 90, "claude", startOffsetSeconds: 30),
            P(90, 80, "pwsh", startOffsetSeconds: 20),
            P(80, 70, "code", startOffsetSeconds: 10),
            P(70, 1, "code", window: true));
        Assert.Equal(new HostChainResult(70, "VS Code", true), HostChain.Resolve(100, table, KnownHosts.Windows));
    }

    [Fact]
    public void APlainConsoleIsTheShellThatOwnsTheWindow()
    {
        var table = Table(P(100, 90, "claude", startOffsetSeconds: 10), P(90, 1, "pwsh", window: true));
        Assert.Equal(new HostChainResult(90, "Console", true), HostChain.Resolve(100, table, KnownHosts.Windows));
    }

    [Fact]
    public void UnknownHostFallsBackToTheFirstWindowedAncestor()
    {
        var table = Table(
            P(100, 90, "claude", startOffsetSeconds: 10),
            P(90, 80, "node", startOffsetSeconds: 5),
            P(80, 1, "myterm", window: true));
        Assert.Equal(new HostChainResult(80, "myterm", false), HostChain.Resolve(100, table, KnownHosts.Windows));
    }

    [Fact]
    public void AKnownHostBeatsAnEarlierUnknownWindowedAncestor()
    {
        var table = Table(
            P(100, 90, "claude", startOffsetSeconds: 20),
            P(90, 80, "helper", window: true, startOffsetSeconds: 10),
            P(80, 1, "windowsterminal", window: true));
        Assert.Equal("Windows Terminal", HostChain.Resolve(100, table, KnownHosts.Windows)!.DisplayName);
    }

    [Fact]
    public void StopsAtTheShellWithoutPickingIt()
    {
        var table = Table(P(100, 90, "claude", startOffsetSeconds: 10), P(90, 1, "explorer", window: true));
        Assert.Null(HostChain.Resolve(100, table, KnownHosts.Windows));
    }

    [Fact]
    public void MissingParentOrMissingStartStopsTheWalk()
    {
        Assert.Null(HostChain.Resolve(100, Table(P(100, 999, "claude")), KnownHosts.Windows));
        Assert.Null(HostChain.Resolve(5, Table(P(100, 1, "claude")), KnownHosts.Windows));
    }

    [Fact]
    public void AParentYoungerThanItsChildIsAReusedPid()
    {
        var table = Table(P(100, 90, "claude"), P(90, 1, "windowsterminal", window: true, startOffsetSeconds: 60));
        Assert.Null(HostChain.Resolve(100, table, KnownHosts.Windows));
    }

    [Fact]
    public void AProcessThatIsItsOwnParentStopsTheWalk()
    {
        Assert.Null(HostChain.Resolve(100, Table(P(100, 100, "claude")), KnownHosts.Windows));
    }

    [Fact]
    public void WalkIsCappedAtTwelveProcesses()
    {
        static IReadOnlyDictionary<int, ProcessRecord> Chain(int nodesBeforeHost)
        {
            var records = new List<ProcessRecord>();
            for (var i = 0; i < nodesBeforeHost; i++)
                records.Add(P(100 + i, 101 + i, "node", startOffsetSeconds: nodesBeforeHost - i));
            records.Add(P(100 + nodesBeforeHost, 1, "windowsterminal", window: true));
            return records.ToDictionary(r => r.Pid);
        }
        Assert.NotNull(HostChain.Resolve(100, Chain(HostChain.MaxDepth - 1), KnownHosts.Windows));
        Assert.Null(HostChain.Resolve(100, Chain(HostChain.MaxDepth), KnownHosts.Windows));
    }

    [Fact]
    public void CatalogsMapNamesAndStopWords()
    {
        Assert.Equal("Terminal", KnownHosts.Mac.DisplayName("com.apple.Terminal"));
        Assert.Equal("VS Code", KnownHosts.Mac.DisplayName("com.microsoft.VSCode"));
        Assert.Equal("JetBrains", KnownHosts.Mac.DisplayName("com.jetbrains.intellij"));
        Assert.Null(KnownHosts.Mac.DisplayName("com.example.other"));
        Assert.True(KnownHosts.Mac.StopsWalk("com.apple.finder"));
        Assert.False(KnownHosts.Mac.StopsWalk("com.apple.Terminal"));
        Assert.Equal("Git Bash", KnownHosts.Windows.DisplayName("mintty"));
        Assert.Equal("JetBrains", KnownHosts.Windows.DisplayName("rider64"));
        Assert.True(KnownHosts.Windows.StopsWalk("explorer"));
        Assert.False(KnownHosts.Windows.StopsWalk("code"));
    }
}

File: mac/Tests/ClaudeToolbarCoreTests/Sessions/HostChainTests.swift
import XCTest
@testable import ClaudeToolbarCore

final class HostChainTests: XCTestCase {
    private let t0 = Date(timeIntervalSince1970: 1_788_868_800)

    // startOffset: larger = started later. Children start after their parents.
    private func p(_ pid: Int, _ parent: Int, _ name: String, window: Bool = false, startOffset: TimeInterval = 0) -> ProcessRecord {
        ProcessRecord(pid: pid, parentPid: parent, name: name, startTime: t0.addingTimeInterval(startOffset), hasWindow: window)
    }

    private func table(_ records: ProcessRecord...) -> [Int: ProcessRecord] {
        Dictionary(uniqueKeysWithValues: records.map { ($0.pid, $0) })
    }

    func testFindsTerminalThreeLevelsUp() {
        let t = table(
            p(100, 90, "node", startOffset: 30),
            p(90, 80, "zsh", startOffset: 20),
            p(80, 70, "login", startOffset: 10),
            p(70, 1, "com.apple.Terminal", window: true))
        XCTAssertEqual(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.mac),
                       HostChainResult(pid: 70, displayName: "Terminal", known: true))
    }

    func testTheClaudeCliIsNotTheDesktopApp() {
        let t = table(p(100, 90, "claude", startOffset: 10), p(90, 1, "windowsterminal", window: true))
        XCTAssertEqual(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.windows)?.displayName, "Windows Terminal")
    }

    func testTheDesktopAppCountsWhenItOwnsAWindow() {
        let t = table(p(100, 50, "claude", startOffset: 10), p(50, 1, "com.anthropic.claudefordesktop", window: true))
        XCTAssertEqual(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.mac),
                       HostChainResult(pid: 50, displayName: "Claude", known: true))
    }

    func testVsCodeHelperWithoutAWindowIsSkippedForItsMainProcess() {
        let t = table(
            p(100, 90, "node", startOffset: 30),
            p(90, 80, "zsh", startOffset: 20),
            p(80, 70, "com.microsoft.VSCode.helper.plugin", startOffset: 10),
            p(70, 1, "com.microsoft.VSCode", window: true))
        XCTAssertEqual(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.mac),
                       HostChainResult(pid: 70, displayName: "VS Code", known: true))
    }

    func testAPlainConsoleIsTheShellThatOwnsTheWindow() {
        let t = table(p(100, 90, "claude", startOffset: 10), p(90, 1, "pwsh", window: true))
        XCTAssertEqual(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.windows),
                       HostChainResult(pid: 90, displayName: "Console", known: true))
    }

    func testUnknownHostFallsBackToTheFirstWindowedAncestor() {
        let t = table(p(100, 90, "node", startOffset: 10), p(90, 80, "zsh", startOffset: 5), p(80, 1, "com.example.myterm", window: true))
        XCTAssertEqual(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.mac),
                       HostChainResult(pid: 80, displayName: "com.example.myterm", known: false))
    }

    func testAKnownHostBeatsAnEarlierUnknownWindowedAncestor() {
        let t = table(
            p(100, 90, "node", startOffset: 20),
            p(90, 80, "com.example.helper", window: true, startOffset: 10),
            p(80, 1, "com.googlecode.iterm2", window: true))
        XCTAssertEqual(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.mac)?.displayName, "iTerm2")
    }

    func testStopsAtTheShellWithoutPickingIt() {
        let t = table(p(100, 90, "node", startOffset: 10), p(90, 1, "com.apple.finder", window: true))
        XCTAssertNil(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.mac))
        let w = table(p(100, 90, "claude", startOffset: 10), p(90, 1, "explorer", window: true))
        XCTAssertNil(HostChain.resolve(startPid: 100, table: w, catalog: KnownHosts.windows))
    }

    func testMissingParentOrMissingStartStopsTheWalk() {
        XCTAssertNil(HostChain.resolve(startPid: 100, table: table(p(100, 999, "node")), catalog: KnownHosts.mac))
        XCTAssertNil(HostChain.resolve(startPid: 5, table: table(p(100, 1, "node")), catalog: KnownHosts.mac))
    }

    func testAParentYoungerThanItsChildIsAReusedPid() {
        let t = table(p(100, 90, "node"), p(90, 1, "com.apple.Terminal", window: true, startOffset: 60))
        XCTAssertNil(HostChain.resolve(startPid: 100, table: t, catalog: KnownHosts.mac))
    }

    func testAProcessThatIsItsOwnParentStopsTheWalk() {
        XCTAssertNil(HostChain.resolve(startPid: 100, table: table(p(100, 100, "node")), catalog: KnownHosts.mac))
    }

    func testWalkIsCappedAtTwelveProcesses() {
        func chain(_ nodesBeforeHost: Int) -> [Int: ProcessRecord] {
            var records: [ProcessRecord] = []
            for i in 0..<nodesBeforeHost {
                records.append(p(100 + i, 101 + i, "node", startOffset: TimeInterval(nodesBeforeHost - i)))
            }
            records.append(p(100 + nodesBeforeHost, 1, "com.apple.Terminal", window: true))
            return Dictionary(uniqueKeysWithValues: records.map { ($0.pid, $0) })
        }
        XCTAssertNotNil(HostChain.resolve(startPid: 100, table: chain(HostChain.maxDepth - 1), catalog: KnownHosts.mac))
        XCTAssertNil(HostChain.resolve(startPid: 100, table: chain(HostChain.maxDepth), catalog: KnownHosts.mac))
    }

    func testCatalogsMapNamesAndStopWords() {
        XCTAssertEqual(KnownHosts.mac.displayName("com.apple.Terminal"), "Terminal")
        XCTAssertEqual(KnownHosts.mac.displayName("com.microsoft.VSCode"), "VS Code")
        XCTAssertEqual(KnownHosts.mac.displayName("com.jetbrains.intellij"), "JetBrains")
        XCTAssertNil(KnownHosts.mac.displayName("com.example.other"))
        XCTAssertTrue(KnownHosts.mac.stopsWalk("com.apple.finder"))
        XCTAssertFalse(KnownHosts.mac.stopsWalk("com.apple.Terminal"))
        XCTAssertEqual(KnownHosts.windows.displayName("mintty"), "Git Bash")
        XCTAssertEqual(KnownHosts.windows.displayName("rider64"), "JetBrains")
        XCTAssertTrue(KnownHosts.windows.stopsWalk("explorer"))
        XCTAssertFalse(KnownHosts.windows.stopsWalk("code"))
    }
}
